File: js/src/chunk-diff.js
// @flow

import {invariant, notNull} from './assert.js';
import {compare} from './compare.js';
import RefValue from './ref-value.js';
import {ValueReader} from './value-store.js';

// RefQueue is a priority queue-like datastructure of RefValue objects, optimised for the RefValue
// diff algorithm. Compared to a general purpose (e.g. heap) priority queue:
//  - offer() has equivalent semantics. O(number of heights), but see NumPQ for why this is fine.
//  - getHeight() instead of peek, because diff only cares about peeking at the height. O(1).
//  - poll() polls all RefValue objects with the largest height, instead of the single highest
//    priority RefValue, because diff works by considering each layer at a time. This means that
//    poll can trivially be O(1).
export class RefQueue {
  _nums: ?NumPQ;
  _refs: Map<number, [RefValue]>;

  constructor() {
    this._refs = new Map();
  }

  offer(...rs: Array<RefValue>) {
    rs.forEach(r => this._offer(r));
  }

  _offer(r: RefValue) {
    const height = r.height;
    const refsAtHeight = this._refs.get(height);

    if (refsAtHeight) {
      refsAtHeight.push(r);
    } else {
      if (this._nums) {
        this._nums.offer(height);
      } else {
        this._nums = new NumPQ(height);
      }

      this._refs.set(height, [r]);
    }
  }

  getHeight(): number {
    const head = this._nums;
    invariant(head, 'cannot get height of empty queue');
    return head.val;
  }

  poll(): [RefValue] {
    const head = this._nums;
    invariant(head, 'cannot poll empty queue');
    this._nums = head.tail;
    const res = notNull(this._refs.get(head.val));
    this._refs.delete(head.val);
    return res;
  }

  isEmpty(): boolean {
    return !this._nums;
  }
}

// NumPQ is a priority queue of RefValue heights implemented as a linked list. This is a reasonable
// implementation (versus a heap, for example) because heights will likely be added close to the
// head, given the way the graph is descended.
class NumPQ {
  val: number;
  tail: ?NumPQ;

  constructor(val: number, tail: ?NumPQ = null) {
    this.val = val;
    this.tail = tail;
  }

  offer(n: number) {
    invariant(n !== this.val, `${n} is already in the queue`);
    if (n > this.val) {
      this.tail = new NumPQ(this.val, this.tail);
      this.val = n;
    } else if (!this.tail) {
      this.tail = new NumPQ(n);
    } else if (n < this.val) {
      this.tail.offer(n);
    }
  }
}

// Computes the number of RefValue objects only reachable from rootA ("onlyA") and only reachable
// from rootB ("onlyB"), in O(onlyA.length + onlyB.length). Returns the tuple [onlyA, onlyB].
export default async function diff(
    vr: ValueReader, rootA: RefValue, rootB: RefValue): Promise<[[RefValue], [RefValue]]> {

  // For each ref in |newRefs|, add to |dest|, then offer its reachable chunks to |reachable|.
  const sync = (newRefs: [RefValue], reachable: RefQueue, dest: [RefValue]): Promise<void> => {
    dest.push(...newRefs);
    return Promise.all(newRefs.map(r => r.targetValue(vr))).then(vs => {
      vs.forEach(v => reachable.offer(...v.chunks));
    });
  };

  const reachableFromA = new RefQueue(), reachableFromB = new RefQueue();
  const onlyInA: [RefValue] = [], onlyInB: [RefValue] = [];

  reachableFromA.offer(rootA);
  reachableFromB.offer(rootB);

  while (!reachableFromA.isEmpty() && !reachableFromB.isEmpty()) {
    const heightA = reachableFromA.getHeight();
    const heightB = reachableFromB.getHeight();
    if (heightA > heightB) {
      const newRefs = pollTo(reachableFromA, heightB);
      await sync(newRefs, reachableFromA, onlyInA);
    } else if (heightB > heightA) {
      const newRefs = pollTo(reachableFromB, heightA);
      await sync(newRefs, reachableFromB, onlyInB);
    } else {
      const [newRefsA, newRefsB] = diffRefValueArrays(pollTo(reachableFromA, heightA - 1),
                                                      pollTo(reachableFromB, heightB - 1));
      await Promise.all([sync(newRefsA, reachableFromA, onlyInA),
                         sync(newRefsB, reachableFromB, onlyInB)]);
    }
  }

  onlyInA.push(...pollTo(reachableFromA, 0));
  onlyInB.push(...pollTo(reachableFromB, 0));

  return [onlyInA, onlyInB];
}

// Returns all RefValue objects in |reachable| that are higher than |floor| (exclusive).
function pollTo(reachable: RefQueue, floor: number): [RefValue] {
  const res = [];
  while (!reachable.isEmpty() && reachable.getHeight() > floor) {
    res.push(...reachable.poll());
  }
  return res;
}

// Sorts |sliceA| and |arrB|, then returns a tuple of [refs only in sliceA, refs only in arrB].
function diffRefValueArrays(sliceA: [RefValue], arrB: [RefValue]): [[RefValue], [RefValue]] {
  const onlyInA: [RefValue] = [], onlyInB: [RefValue] = [];
  let idxA = 0, idxB = 0;

  sliceA.sort(compare);
  arrB.sort(compare);

  while (idxA < sliceA.length && idxB < arrB.length) {
    const refA = sliceA[idxA], refB = arrB[idxB];
    switch (compare(refA, refB)) {
      case -1:
        onlyInA.push(refA);
        idxA++;
        break;
      case 0:
        idxA++;
        idxB++;
        break;
      case 1:
        onlyInB.push(refB);
        idxB++;
        break;
    }
  }

  onlyInA.push(...sliceA.slice(idxA));
  onlyInB.push(...arrB.slice(idxB));

  return [onlyInA, onlyInB];
}

File: types/chunk_diff.go
package types

// TODO: rename file to chunks_diff/chunks_diff_test.go.

import (
	"container/heap"
	"sort"
)

// refHeap implements heap.Interface (which includes sort.Interface) as a height based priority queue.
type refHeap []Ref

func (h refHeap) Head() Ref {
	return h[0]
}

func (h refHeap) Len() int {
	return len(h)
}

func (h refHeap) Less(i, j int) bool {
	// > because we want the larger heights to be at the start of the queue.
	return h[i].Height() > h[j].Height()
}

func (h refHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *refHeap) Push(r interface{}) {
	*h = append(*h, r.(Ref))
}

func (h *refHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// refSlice implements sort.Interface to order by target ref.
type refSlice []Ref

func (s refSlice) Len() int {
	return len(s)
}

func (s refSlice) Less(i, j int) bool {
	return s[i].TargetRef().Less(s[j].TargetRef())
}

func (s refSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// TODO: option to exclude one of the result tuples (e.g. might only care about chunks in rootA that aren't in rootB).
// TODO: support concurrency. The JS implementation will use Promises; can we use Go to implement something similar?
func ChunksDiff(vrA, vrB ValueReader, rootA, rootB Ref) ([]Ref, []Ref) {

	// TODO: comment.
	commit := func(vr ValueReader, onlyIn refSlice, reachable *refHeap, r Ref) refSlice {
		if chunks := r.TargetValue(vr).Chunks(); chunks != nil {
			for _, chunk := range chunks {
				heap.Push(reachable, chunk)
			}
		}
		return append(onlyIn, r)
	}

	// TODO: comment.
	syncTo := func(vr ValueReader, onlyIn refSlice, reachable *refHeap, height uint64) refSlice {
		for reachable.Len() > 0 && reachable.Head().Height() > height {
			onlyIn = commit(vr, onlyIn, reachable, heap.Pop(reachable).(Ref))
		}
		return onlyIn
	}

	// TODO: comment.
	popTo := func(reachable *refHeap, height uint64) (res refSlice) {
		for reachable.Len() > 0 && reachable.Head().Height() > height {
			res = append(res, heap.Pop(reachable).(Ref))
		}
		return
	}

	// TODO: comment.
	diffRefSlices := func(sliceA, sliceB refSlice) (refSlice, refSlice) {
		onlyInA, onlyInB := refSlice{}, refSlice{}
		idxA, idxB := uint64(0), uint64(0)

		sort.Sort(sliceA)
		sort.Sort(sliceB)

		for idxA < uint64(len(sliceA)) && idxB < uint64(len(sliceB)) {
			refA, refB := sliceA[idxA], sliceB[idxB]
			switch {
			case refA.TargetRef() == refB.TargetRef():
				idxA++
				idxB++
			case refA.TargetRef().Less(refB.TargetRef()):
				onlyInA = append(onlyInA, refA)
				idxA++
			default:
				onlyInB = append(onlyInB, refB)
				idxB++
			}
		}

		onlyInA = append(onlyInA, sliceA[idxA:]...)
		onlyInB = append(onlyInB, sliceB[idxB:]...)
		return onlyInA, onlyInB
	}

	onlyInA, onlyInB := refSlice{}, refSlice{}
	reachableFromA, reachableFromB := refHeap{}, refHeap{}

	heap.Init(&reachableFromA)
	heap.Init(&reachableFromB)
	heap.Push(&reachableFromA, rootA)
	heap.Push(&reachableFromB, rootB)

	for len(reachableFromA) > 0 && len(reachableFromB) > 0 {
		heightA := reachableFromA[0].Height()
		heightB := reachableFromB[0].Height()

		if heightA > heightB {
			onlyInA = syncTo(vrA, onlyInA, &reachableFromA, heightB)
		} else if heightB > heightA {
			onlyInB = syncTo(vrB, onlyInB, &reachableFromB, heightA)
		} else {
			newRefsA, newRefsB := diffRefSlices(popTo(&reachableFromA, heightA-1), popTo(&reachableFromB, heightB-1))
			for _, r := range newRefsA {
				onlyInA = commit(vrA, onlyInA, &reachableFromA, r)
			}
			for _, r := range newRefsB {
				onlyInB = commit(vrB, onlyInB, &reachableFromB, r)
			}
		}
	}

	onlyInA = append(onlyInA, reachableFromA...)
	onlyInB = append(onlyInB, reachableFromB...)
	return []Ref(onlyInA), []Ref(onlyInB)
}
